presto-main/src/main/java/com/facebook/presto/metadata/Signature.java

```diff
@@ -328,15 +328,22 @@ private static boolean matchAndBind(Map<String, Type> boundParameters, Map<Strin
     {
         // If this parameter is already bound, then match (with coercion)
         if (boundParameters.containsKey(parameter.getBase())) {
-            checkArgument(parameter.getParameters().isEmpty(), "Unexpected parameteric type");
+            checkArgument(parameter.getParameters().isEmpty(), "Unexpected parametric type");
             if (allowCoercion) {
                 if (canCoerce(type, boundParameters.get(parameter.getBase()))) {
                     return true;
                 }
                 else if (canCoerce(boundParameters.get(parameter.getBase()), type) && typeParameters.get(parameter.getBase()).canBind(type)) {
-                    // Broaden the binding
+                    // Try to coerce current binding to new candidate
                     boundParameters.put(parameter.getBase(), type);
                     return true;
+                } else {
+                    // Try to use common super type of current binding and candidate
```

Comment:
this probably should only be considered only for the same base types and only for some of them (e.g. decimal). That is why I like the idea from *unification* patch by haozun: *After this commit, coercibility between types is decoupled into 2 parts: compatibility between their based types and between their type parameters.*.
That also means that method: *getCommonSuperType* should probably be more specific or split (depending on other usages), e.g. getCoercionSuperType.